### PR TITLE
Fix SourcePos of Let

### PR DESCRIPTION
**Pull request by HoshinoTented**
image

### CODE REVIEW

File: base/src/main/java/org/aya/concrete/Expr.java [hunk 1]
```diff
@@ -708,15 +708,36 @@ public static Expr.Array newGenerator(
...
   record Let(
+    @NotNull SourcePos sourcePos,
     @NotNull Let.Bind bind,
     @NotNull Expr body
   ) implements Expr {
+    public Let {
+      if (body instanceof Let let) {
+        sourcePos = let.sourcePos();
+      }
+    }
```
Comment (imkiva) [inline base/src/main/java/org/aya/concrete/Expr.java:739]: The syntax location should not be the same as the error reporting location. The correct `sourcePos` of a `Let` should always be `sourcePosOf(GenericNode)` in the producer. Or some location-based features may not work freely for `Let`

Comment (HoshinoTented) [reply base/src/main/java/org/aya/concrete/Expr.java:739]: So what do we expect? The whole let-expression or the body?

Comment (ice1000) [reply base/src/main/java/org/aya/concrete/Expr.java:739]: 0. Let should have whole SourcePos
1. LetBind can have another SourcePos and we use that for error report

Comment (HoshinoTented) [reply base/src/main/java/org/aya/concrete/Expr.java]: So: For all `Let`s in the same group, they share the same source pos -- the whole source pos

Comment (HoshinoTented) [reply base/src/main/java/org/aya/concrete/Expr.java]: I think it is good to have a way to obtain the source pos of an expr that determines the type of the whole expr (like the last `DoBind` of a do-notation, the body of `Let`)

Comment (imkiva) [reply base/src/main/java/org/aya/concrete/Expr.java]: > So: For all `Let`s in the same group, they share the same source pos -- the whole source pos

No

Comment (imkiva) [reply base/src/main/java/org/aya/concrete/Expr.java]: > I think it is good to have a way to obtain the source pos of an expr that determines the type of the whole expr (like the last `DoBind` of a do-notation, the body of `Let`)

Yes

**Pull request merged**